File: src/main/java/kr/pe/ecmaxp/openpie/Call.java
package kr.pe.ecmaxp.openpie;

import li.cil.oc.api.machine.LimitReachedException;
import li.cil.oc.api.machine.Machine;

import java.util.Arrays;

public class Call
{
    public final String component;
    public final String function;
    public final Object[] args;

    public Call(String component, String function, Object... args)
    {
        this.component = component;
        this.function = function;
        this.args = args;
    }

    public static Call FromObjectArray(Object[] array)
    {
        if (array.length < 2)
            return null;

        Object[] args = new Object[array.length - 2];
        System.arraycopy(array, 2, args, 0, args.length);

        return new Call((String) array[0], (String) array[1], args);
    }

    @Override
    public String toString()
    {
        return "Call{" +
                "component='" + component + '\'' +
                ", function='" + function + '\'' +
                ", args=" + Arrays.toString(args) +
                '}';
    }

    public Result invoke(Machine machine) throws LimitReachedException
    {
        try
        {
            Object[] result = machine.invoke(component, function, args);
            return new Result(this, result);
        }
        catch (LimitReachedException e)
        {
            throw e;
        }
        catch (Error e)
        {
            return new Result(this, e);
        }
        catch (Exception e)
        {
            return new Result(this, e);
        }
    }
}

File: src/main/java/kr/pe/ecmaxp/openpie/arch/types/Call.kt
package kr.pe.ecmaxp.openpie.arch.types

import li.cil.oc.api.machine.LimitReachedException
import li.cil.oc.api.machine.Machine
import java.util.*

class Call(val component: String, val function: String, vararg args: Any?) {
    val args: Array<*> = args

    override fun toString(): String {
        return "Call{" +
                "component='" + component + '\''.toString() +
                ", function='" + function + '\''.toString() +
                ", args=" + Arrays.toString(args) +
                '}'.toString()
    }

    @Throws(LimitReachedException::class)
    operator fun invoke(machine: Machine): Result {
        try {
            val result = machine.invoke(component, function, args)
            return Result(this, result)
        } catch (e: LimitReachedException) {
            throw e
        } catch (e: Error) {
            return Result(this, e)
        } catch (e: Exception) {
            return Result(this, e)
        }
    }

    companion object {
        fun FromObjectArray(array: Array<*>): Call? {
            if (array.size < 2)
                return null

            val args = arrayOfNulls<Any?>(array.size - 2)
            System.arraycopy(array, 2, args, 0, args.size)

            return Call(array[0] as String, array[1] as String, *args)
        }
    }
}
